File: python_package/flourishcharts/flourish.py
#!/usr/bin/env python
# coding: utf-8

# Copyright (c) Zoe Meers.
# Distributed under the terms of the Modified BSD License.

"""
Create Flourish object.
"""
import os
import importlib.metadata
import pathlib
import requests
import anywidget
import traitlets

try:
    __version__ = importlib.metadata.version("flourishcharts")
except importlib.metadata.PackageNotFoundError:
    __version__ = "unknown"

from . import bindings, details
from ._utils import load_internal_templates


class Flourish(
    anywidget.AnyWidget,
    details.DetailsMixin,
    bindings.DataBindingsMixin,
):
    """Create a Flourish graph.

    :param chart_type: A chart type string for a graph in Flourish's library. Preferred usage compared to template_id and template_version.
    :type chart_type: str, optional
    :param base_visualisation_id: Optional - provide the numerical ID of an existing public Flourish visualisation to reference it's configuration
    :type base_visualisation_id: str, optional
    :param chart_description: A screen-reader description. A text alternative to the visual content that will only be visible to screen-readers, e.g. “The line chart shows China consistently higher than the other countries since 1990”. Do not replicate your title, since that will also be read by screenreaders., defaults to None
    :type chart_description: str, optional
    :param template_id: A template ID, as set by Flourish, found at https://app.flourish.studio/@flourish. Use if you cannot use `chart_type` or `base_visualisation_id`, Defaults to None
    :type template_id: str, optional
    :param template_version_number: A template version number, as set by Flourish, found at https://app.flourish.studio/@flourish. Use if you cannot use `chart_type` or `base_visualisation_id`, Defaults to None
    :type template_version_number: str, optional
    :param width: Width of the graph, defaults to None
    :type width: int, optional
    :param height: Height of the graph, defaults to None
    :type height: int, optional
    :param api_key: Flourish API key, defaults to None
    :type api_key: str, optional
    """

    _esm = pathlib.Path(__file__).parent / "static" / "widget.js"
    _css = pathlib.Path(__file__).parent / "static" / "widget.css"
    _model_data = traitlets.Dict().tag(sync=True)

    def __init__(
        self,
        chart_type: str = None,
        chart_description: str = None,
        base_visualisation_id: str = None,
        template_id: str = None,
        template_version_number: str = None,
        width: int = None,
        height: int = None,
        api_key: str = None,
    ):
        super().__init__()
        self.chart_description = chart_description
        self.chart_type = chart_type
        self.base_visualisation_id = base_visualisation_id
        self.width = width
        self.height = height
        self.state = {}

        # Check for API key
        if api_key is not None:
            self.api_key = api_key
        elif os.environ.get("FLOURISH_API_KEY") is not None:
            self.api_key = os.environ.get("FLOURISH_API_KEY")
        else:
            raise Exception(
                "Flourish API key not found. Please grab a key from Flourish settings "
                "https://app.flourish.studio/settings and add them to your .envrc file."
            )

        # match to a template
        _internal_templates = load_internal_templates()
        if chart_type is not None:
            matched_templates = _internal_templates[
                _internal_templates.chart_type == chart_type
            ].to_dict(orient="records")

            if not matched_templates and chart_type:
                raise Exception(
                    f"The chart type [{chart_type}] is not an exact match with Flourish's available graphs. "
                    f"Find available chart types in `flourishcharts.flourish_api_documentation`, or specify a template ID and version number, or the base visualisation ID of an existing published Flourish graph (if available)."
                )
            if matched_templates and chart_type:
                matched_template = matched_templates[
                    0
                ]  # arbitrarily take the first matching template

                self.template_id = matched_template["template_id"]
                self.template_version = matched_template["template_version"]
                if matched_template["template_id"] == "@flourish/hierarchy":
                    self.state = {"hierarchy_layout": chart_type}
                if matched_template["template_id"] == "@flourish/line-bar-pie":
                    self.state = {"chart_type": chart_type}

        if base_visualisation_id is not None:
            base_viz_object = f"https://public.flourish.studio/visualisation/{base_visualisation_id}/visualisation-object.json"
            base_viz_data = requests.get(base_viz_object)
            if base_viz_data.status_code == 403:
                raise Exception(
                    f"Access denied accessing base visualisation {base_viz_object} - please ensure it is published for public use"
                )
            else:
                base_viz_dict = base_viz_data.json()
                self.template_id = base_viz_dict["template"]
                self.template_version = base_viz_dict["version"]
                self.base_bindings = base_viz_dict["bindings"]
                self.base_data = base_viz_dict["data"]
                self.base_state = base_viz_dict["state"]
                if base_viz_dict["metadata"]:
                    self.base_metadata = base_viz_dict["metadata"]
                else:
                    self.base_metadata = {}
        else:
            self.base_bindings = {}
            self.base_data = {}
            self.base_state = {}
            self.base_metadata = {}

        if chart_type is None and base_visualisation_id is None and template_id is None:
            raise Exception(
                "Error: please enter a chart type, template ID and version number, or the base visualisation ID of an existing published Flourish graph, if available."
            )
        if chart_type is None and base_visualisation_id is None:
            if not template_id and template_version_number is None:
                raise Exception("Error: please specify a template version number.")
            if template_id is None and not template_version_number:
                raise Exception("Error: please specify a template ID.")
            if not template_id and not template_version_number:
                self.template_id = template_id
                self.template_version = template_version_number

        if chart_type is not None and chart_type.startswith(("map", "globe")):
            raise Exception(
                f"The chart type [{chart_type}] is a map template and is currently not supported in this module."
            )

        if base_visualisation_id is not None:
            self._model_data = {
                "state": self.base_state,
                "data": self.base_data,
                "bindings": self.base_bindings,
                "metadata": self.base_metadata,
                "chart_description": self.chart_description,
                "template": self.template_id,
                "version": self.template_version,
                "api_key": self.api_key,
                "width": self.width,
                "height": self.height,
                "base_visualisation_data_format": "object"
            }
        if base_visualisation_id is None:
            self._model_data = {
                "state": self.state,
                "base_visualisation_id": self.base_visualisation_id,
                "data": {"data": list()},  # set an empty list so the base graph builds
                "chart_description": self.chart_description,
                "template": self.template_id,
                "version": self.template_version,
                "api_key": self.api_key,
                "width": self.width,
                "height": self.height
            }

    def __repr__(self):
        return f"Flourish(template_id={self.template_id}, template_version={self.template_version}"

File: R_package/R/flourish.R
#' Plot a Flourish graph.
#'
#' Create a graph using Flourish's API (https://developers.flourish.studio/api/introduction/).
#' @name flourish
#' @param chart_type Required - a chart type string for each graph in Flourish's library. Preferred usage compared to `template_id` and `template_version`.
#' @param base_visualisation_id Optional - provide the numerical ID of an existing public (published) Flourish visualisation to reference it's configuration
#' @param template_id Optional - the Flourish template ID, found at https://app.flourish.studio/@flourish. Use if you cannot use `chart_type`.
#' @param template_version_number Optional - the template's version number. Also found at https://app.flourish.studio/@flourish. Use if you cannot use `chart_type`.
#' @param chart_description Optional - a screen-reader description. A text alternative to the visual content that will only be visible to screen-readers, e.g. “The line chart shows China consistently higher than the other countries since 1990”. Do no replicate your title, since that will also be read by screenreaders.
#' @param api_key Required - our API key which can be generated within your profile account. Defaults to Sys.getenv("FLOURISH_API_KEY").
#' @return A Flourish chart
#' @examples 
#' try(
#'   flourish(
#'     chart_type = "scatter", api_key = Sys.getenv("FLOURISH_API_KEY")
#'   ) |> 
#'   bind_scatter_data(
#'     data = subset(gapminder, year == 2007), 
#'     y = "gdpPercap", x = "lifeExp"
#'   )
#'  )
#' @export

flourish <- function(chart_type = NULL,
                     chart_description = NULL,
                     base_visualisation_id = NULL,
                     template_id = NULL,
                     template_version_number = NULL,
                     api_key = Sys.getenv("FLOURISH_API_KEY")) {
  api_version <- "5.1.0"
  templates_chart_type_data <- templates_private[templates_private$chart_type == chart_type, ]
  if (!is.null(base_visualisation_id)) {
    base_visualisation_data_format <- "object"
    # download the base visualisation's configuration to get template_id/version to setup the iframe
    base_visualisation_url <- sprintf("https://public.flourish.studio/visualisation/%d/visualisation-object.json", base_visualisation_id)
    resp <- httr::GET(base_visualisation_url)
    if (resp$status_code == 403) {
      stop(paste("Access denied accessing base visualisation ", base_visualisation_url, " - please ensure it is published for public use"))
    }
    jsonRespParsed <- httr::content(resp, as = "parsed")
    template_id <- jsonRespParsed$template
    template_version <- jsonRespParsed$version
    # The Flourish Live API now supports array of objects, so let's convert them as it's just easier to work with data frames in R:
    # loop through datasets (there might be multiple, depending on template)
    for (i in seq_along(jsonRespParsed$data)) {
      # Convert the list of lists to a data frame
      df <- as.data.frame(do.call(rbind, jsonRespParsed$data[[i]]))
      # Assign the modified data frame back to the original list
      jsonRespParsed$data[[i]] <- df
    }
    for (i in seq_along(jsonRespParsed$data)) {
      list_element <- jsonRespParsed$metadata[[i]]
      for (col_name in names(jsonRespParsed$data[[i]])) {
        if (list_element[[col_name]]$type == "string") {
          jsonRespParsed$data[[i]][[col_name]][jsonRespParsed$data[[i]][[col_name]] == "NA"] <- ""
          jsonRespParsed$data[[i]][[col_name]] <- as.character(jsonRespParsed$data[[i]][[col_name]])
        } else if (list_element[[col_name]]$type == "number") {
          jsonRespParsed$data[[i]][[col_name]][jsonRespParsed$data[[i]][[col_name]] == "NA"] <- ""
          jsonRespParsed$data[[i]][[col_name]] <- suppressWarnings(as.numeric(jsonRespParsed$data[[i]][[col_name]]))
        } else if (list_element[[col_name]]$type == "datetime") {
          jsonRespParsed$data[[i]][[col_name]][jsonRespParsed$data[[i]][[col_name]] == "NA"] <- ""
          jsonRespParsed$data[[i]][[col_name]] <- as.character(jsonRespParsed$data[[i]][[col_name]])
        }
      }
    }
    if (jsonRespParsed$template %in% c("@flourish/line-bar-pie", "@flourish/hierarchy")) {
      base_viz_chart_type <- base_coalesce_chart_type(jsonRespParsed$template, jsonRespParsed$state$chart_type, jsonRespParsed$state$hierarchy_layout)
    } else {
      base_viz_chart_type <- chart_type
    }
    base_bindings <- jsonRespParsed$bindings
    base_data <- jsonRespParsed$data
    base_state <- jsonRespParsed$state
    base_metadata <- jsonRespParsed$metadata
  } else {
    base_bindings <- NULL
    base_data <- NULL
    base_state <- NULL
    base_metadata <- NULL
    base_visualisation_data_format <- NULL
  }

  if (!is.null(chart_type) && nrow(templates_chart_type_data) > 0) {
    template_id <- templates_chart_type_data$template_id
    template_version <- templates_chart_type_data$template_version
    base_viz_chart_type <- templates_chart_type_data$chart_type
    if (!is.null(base_visualisation_id)) {
      base_viz_chart_type <- NULL
    }
  } else if (is.null(chart_type) && is.null(base_visualisation_id) && is.null(template_id)) {
    stop("Error: please enter a chart type, template ID and version number, or the base visualisation ID of an existing published Flourish graph, if available.")
  } else if (!is.null(base_visualisation_id)) {
    template_id <- jsonRespParsed$template
    template_version <- jsonRespParsed$version
  } else if (is.null(chart_type) && !is.null(template_id) && !is.null(template_version_number)) {
    template_id <- template_id
    template_version <- template_version_number
  } else if (!is.null(chart_type) && !is.null(template_id) && !is.null(template_version_number)) {
    stop("Error: specify only a chart type or a template ID and version number.")
  } else if (!is.null(template_id) && is.null(template_version_number)) {
    stop("Error: please specify a template version number.")
  } else if (is.null(template_id) && !is.null(template_version_number)) {
    stop("Error: please specify a template ID or a chart type.")
  } else if (!is.null(chart_type) && nrow(templates_chart_type_data) < 1) {
    stop("Error: your chart type string is not an exact match with Flourish's available graphs.")
  } else if (grepl("(^map_.*$)|(^globe_.*)", chart_type)) {
    stop("Error: we do not currently support visualizing map templates via the chart type parameter.")
  }

  # generate an ID for the div
  random_string <- c(
    sample(LETTERS, 5, replace = TRUE),
    sample(0:9, 9, replace = TRUE),
    sample(LETTERS, 7, replace = TRUE)
  )
  random_string_value <- paste0(random_string, collapse = "")

  elementId <- sprintf(
    "htmlwidget-%s",
    random_string_value
  )
  chartId <- sprintf(
    "chart-%s",
    random_string_value
  )
  if (isTRUE(knitr::is_html_output())) {
    iframe_src <- paste0("https://flourish-api.com/api/v1/live/template?api_key=", api_key, "&template=", template_id, "&api_version=", api_version, "&version=", template_version)
    iframe_height <- "575px"
  } else {
    iframe_src <- paste0("https://flourish-api.com/api/v1/live/template?api_key=", api_key, "&template=", template_id, "&api_version=", api_version, "&version=", template_version)
    iframe_height <- "100%"
  }

  # forward options using x
  x <- list(
    chart_id = chartId,
    template = template_id,
    version = template_version,
    api_key = api_key,
    chart_type = base_viz_chart_type,
    state = c(flourish_default_layout(chart_description),
      chart_type = base_viz_chart_type,
      hierarchy_layout = base_viz_chart_type
    ),
    elementId = paste0("#", elementId),
    data = NULL,
    base_visualisation_id = base_visualisation_id,
    base_state = dplyr::coalesce(base_state, flourish_default_layout(chart_description)),
    base_bindings = base_bindings,
    base_data = base_data,
    base_metadata = base_metadata,
    base_visualisation_data_format = base_visualisation_data_format
  )

  # create widget
  htmlwidgets::prependContent(
    htmlwidgets::createWidget(
      name = "flourish-live",
      elementId = elementId,
      x,
      package = "flourishcharts",
      sizingPolicy = htmlwidgets::sizingPolicy(
        viewer.padding = 0,
        viewer.fill = TRUE,
        browser.fill = TRUE,
        viewer.defaultWidth = "100%",
        knitr.defaultWidth = "100%"
      )
    ),
    htmltools::tags$iframe(
      id = chartId,
      width = "100%",
      height = iframe_height,
      seamless = "seamless",
      scrolling = "no",
      frameborder = "2",
      title = "Interactive or visual content",
      sandbox = "allow-same-origin allow-forms allow-scripts allow-downloads allow-popups allow-popups-to-escape-sandbox allow-top-navigation-by-user-activation",
      src = iframe_src
    )
  )
}
